edgedb/options.py

```diff
@@ -0,0 +1,291 @@
+import abc
+import enum
+import random
+from collections import namedtuple
+
+from . import errors
+
+
+_RetryRule = namedtuple("_RetryRule", ["attempts", "backoff"])
+
+
+def default_backoff(attempt):
+    return (2 ** attempt) * 0.1 + random.randrange(100) * 0.001
+
+
+class RetryCondition:
+    """Specific condition to retry on for fine-grained control"""
+    SerializationError = enum.auto()
+    Deadlock = enum.auto()
+    NetworkError = enum.auto()
+
+
+class IsolationLevel:
+    """Isolation level for transaction"""
+    Serializable = "SERIALIZABLE"
+    RepeatableRead = "REPEATABLE READ"
+
+
+class RetryOptions:
+    """An immutable class that contains rules for `retrying_transaction()`"""
+    __slots__ = ['_default', '_overrides']
+
+    def __init__(self, attempts: int, backoff):
+        self._default = _RetryRule(attempts, backoff)
+        self._overrides = None
+
+    def with_rule(self, condition, attempts=None, backoff=None):
+        default = self._default
+        overrides = self._overrides
+        if overrides is None:
+            overrides = {}
+        else:
+            overrides = overrides.copy()
+        overrides[condition] = _RetryRule(
+            default.attempts if attempts is None else attempts,
+            default.backoff if backoff is None else backoff,
+        )
+        result = RetryOptions.__new__(RetryOptions)
+        result._default = default
+        result._overrides = overrides
+        return result
+
+    @classmethod
+    def defaults(cls):
+        return cls(
+            attempts=3,
```

Comment:
Does attempts mean that 3 total attempts will be made or that 3 additional attempts will be made after an initial attempt?

Reply:
3 in total. So one initial and two additional in case of error.